### PR TITLE
Config File and StorageProxy Abstractions

### PR DESCRIPTION
**Pull request by Kefaun2601**
## What changes were proposed in this pull request?

Changes to sparkler-default.yaml configuration:
- Abstracted out crawldb.uri into crawldb.backend, solr.uri, and elasticsearch.uri

Added StorageProxy and StorageProxyFactory:
- WIP Abstracting out Solr implementation into the SolrProxy wrapper
- StorageProxyFactory for returning the correct StorageProxy (SolrProxy or ElasticsearchProxy) based on the yaml config file

**Is this related to an already existing issue on sparkler?**  
Related to #211 
Related to #218 

**Will it close an existing issue?**  
Does not close an issue yet.

### How was this patch tested?

This patch was tested by running "mvn clean package" within the "sparkler-core" directory. The tests currently pass, and it builds successfully.

### CODE REVIEW

File: sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/CrawlDbRDD.scala [hunk 1]
```diff
@@ -68,8 +68,8 @@ class CrawlDbRDD(sc: SparkContext,
     qry.set("group.field", groupBy)
     qry.set("group.limit", 0)
     qry.setRows(maxGroups)
-    val proxy = job.newCrawlDbSolrClient()
-    val solr = proxy.crawlDb
+    val proxy = job.newStorageProxy()
+    val solr = proxy.getClient()
```
Comment (lewismc) [inline sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/CrawlDbRDD.scala:72]: Change variable name from `solr` to `client`

File: sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/MemexCrawlDbRDD.scala [hunk 2]
```diff
@@ -47,8 +47,8 @@ class MemexCrawlDbRDD(sc: SparkContext,
     qry.set("group.field", Constants.solr.PARENT)
     qry.set("group.limit", 0)
     qry.setRows(maxGroups)
-    val proxy = job.newCrawlDbSolrClient()
-    val solr = proxy.crawlDb
+    val proxy = job.newStorageProxy()
+    val solr = proxy.getClient()
```
Comment (lewismc) [inline sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/MemexCrawlDbRDD.scala:51]: Change variable name from `solr` to `client`

File: sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/MemexDeepCrawlRDD.scala [hunk 3]
```diff
@@ -57,8 +57,8 @@ class MemexDeepCrawlDbRDD(sc: SparkContext,
     qry.set("group.field", Constants.solr.PARENT)
     qry.set("group.limit", 0)
     qry.setRows(maxGroups)
-    val proxy = job.newCrawlDbSolrClient()
-    val solr = proxy.crawlDb
+    val proxy = job.newStorageProxy()
+    val solr = proxy.getClient()
```
Comment (lewismc) [inline sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/MemexDeepCrawlRDD.scala:61]: Change variable name from `solr` to `client`

File: sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/Crawler.scala [hunk 4]
```diff
@@ -61,7 +60,7 @@ class Crawler extends CliTool {
 
   @Option(name = "-cdb", aliases = Array("--crawldb"),
     usage = "Crawl DB URI.")
-  var sparkSolr: String = sparklerConf.get(Constants.key.CRAWLDB).asInstanceOf[String]
+  var sparkSolr: String = sparklerConf.getDatabaseURI()
```
Comment (lewismc) [inline sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/Crawler.scala:63]: Change variable name from `sparkSolr` to `sparkStorage`

File: sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/Crawler.scala [hunk 5]
```diff
@@ -171,7 +170,7 @@ class Crawler extends CliTool {
     //STEP : Initialize environment
     init()
 
-    val solrc = this.job.newCrawlDbSolrClient()
+    val solrc = this.job.newStorageProxy()
```
Comment (lewismc) [inline sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/Crawler.scala:173]: Change variable name from `solrc` to `sProxy` or `storageProxy`

File: sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/service/Injector.scala [hunk 6]
```diff
@@ -66,7 +66,7 @@ class Injector extends CliTool {
 
   @Option(name = "-cdb", aliases = Array("--crawldb"),
     usage = "Crawdb URI.")
-  var sparkSolr: String = conf.get(Constants.key.CRAWLDB).asInstanceOf[String]
+  var sparkSolr: String = conf.getDatabaseURI()
```
Comment (lewismc) [inline sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/service/Injector.scala:69]: Change variable name from `sparkSolr` to `sparkStorage`

File: sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/service/Injector.scala [hunk 7]
```diff
@@ -109,10 +109,10 @@ class Injector extends CliTool {
     })
     LOG.info("Injecting {} seeds", seeds.size())
 
-    val solrClient = job.newCrawlDbSolrClient()
-    solrClient.addResources(seeds.iterator())
-    solrClient.commitCrawlDb()
-    solrClient.close()
+    val dbClient = job.newStorageProxy()
```
Comment (lewismc) [inline sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/service/Injector.scala:112]: Change variable name from `dbClient` to `sProxy` or `storageProxy`

File: sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/storage/StorageProxy.scala [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package edu.usc.irds.sparkler.storage
+
+/**
+  *
+  * @since 3/2/2021
```
Comment (lewismc) [inline sparkler-core/sparkler-app/src/main/scala/edu/usc/irds/sparkler/storage/StorageProxy.scala:22]: This should not be a date it should be a version release. You can find this in `pom.xml`

**Review by lewismc — CHANGES_REQUESTED** [review]
This looks REALLY good so far. I've made some trivial comments to get you in the right direction. Thank you for working on this guys and posting your pull requests early for review. It makes a huge difference.
Please continue to update this PR and I will review incrementally.

**Comment by lewismc** [comment]
This is looking better @kyan2601 thanks.

**Comment by Kefaun2601** [comment]
Confirmed that I can run a crawl job and view the data on the Banana dashboard. Did not notice any regressions in functionality that may have arose from the implemented Solr abstractions.

Commands run (as per the [repo README](https://github.com/USCDataScience/sparkler/blob/master/README.md)):
<pre>
# Start in the sparkler-core directory
cd sparkler-core

# Run script to start docker container and forward ports to host
bash ./bin/dockler.sh

# Inject seed urls
/data/sparkler/bin/sparkler.sh inject -id 1 -su 'http://www.bbc.com/news'

# Start the crawl job
/data/sparkler/bin/sparkler.sh crawl -id 1 -tn 100 -i 2
</pre>

Access the Banana dashboard at [http://localhost:8983/banana/](http://localhost:8983/banana/) to see the data.

**Comment by lewismc** [comment]
HJi @Kefaun2601 this PR and branch has a conflict which must be resolved. Once that's done, please tag me and I will test it out. Thank you

**Comment by Kefaun2601** [comment]
@lewismc Resolved the merge conflicts. Could you please review it? Thanks!

**Comment by Kefaun2601** [comment]
Added documentation for specifying the storage engine to use in the config file (sparkler-default.yaml):
[https://github.com/USCDataScience/sparkler/wiki/Specifying-CrawlDB-in-Config](https://github.com/USCDataScience/sparkler/wiki/Specifying-CrawlDB-in-Config)

Documentation on the StorageProxyFactory abstraction will be coming.

**Comment by Kefaun2601** [comment]
Added documentation for the StorageProxyFactory abstraction. Will build on this documentation as we expand the factory.

[https://github.com/USCDataScience/sparkler/wiki/StorageProxyFactory-Abstraction](https://github.com/USCDataScience/sparkler/wiki/StorageProxyFactory-Abstraction)

**Comment by lewismc** [comment]
Also confirmed that this DOES NOT introduce a regression. Thank you @felixloesing and @kyan2601

**Pull request merged**